src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1337,28 +1419,41 @@ public TValue AddOrUpdate(TKey key, TValue addValue, Func<TKey, TValue, TValue>
                 ThrowHelper.ThrowArgumentNullException(nameof(updateValueFactory));
             }
 
-            IEqualityComparer<TKey>? comparer = _comparer;
-            int hashcode = comparer is null ? key.GetHashCode() : comparer.GetHashCode(key);
+            Tables tables = _tables;
+
+            IEqualityComparer<TKey>? comparer = tables._comparer;
+            int hashcode = GetHashCode(comparer, key);
 
             while (true)
             {
-                if (TryGetValueInternal(key, hashcode, out TValue? oldValue))
+                if (TryGetValueInternal(tables, key, hashcode, out TValue? oldValue))
                 {
                     // key exists, try to update
                     TValue newValue = updateValueFactory(key, oldValue);
-                    if (TryUpdateInternal(key, hashcode, newValue, oldValue))
+                    if (TryUpdateInternal(tables, key, hashcode, newValue, oldValue))
                     {
                         return newValue;
                     }
                 }
                 else
                 {
                     // key doesn't exist, try to add
-                    if (TryAddInternal(key, hashcode, addValue, updateIfExists: false, acquireLock: true, out TValue resultingValue))
+                    if (TryAddInternal(tables, key, hashcode, addValue, updateIfExists: false, acquireLock: true, out TValue resultingValue))
                     {
                         return resultingValue;
                     }
                 }
+
+                Tables newTables = _tables;
+                if (tables != newTables)
+                {
+                    tables = newTables;
```

Comment:
Why there is local variable `Tables newTables`? Wouldn't it be enough to check and assign the against `_tables`?

Reply:
Do you mean why isn't it:
```C#
if (tables != _tables)
{
    tables = _tables;
    ...
}
```
?  If so, to avoid an unnecessary second volatile read.

Reply:
Yup, that is what I mean. And I was confused because I saw this pattern in Try(Remove|Update|Add)Internal methods above. Or it cant be updated there?

Reply:
I can make them consistent.  We already expect this to be a rare occurrence, so it's not really a big deal either way.